Batching job-submit tasks to avoid blocking Stdout/err pipes

wxtim opened the pull request:
Address part of #2857 
Temporary PR to check fix on codacy.

wxtim commented:
1. I created a new function as a sub-function of `taskJobManager.submit_task_jobs` called `list_chopper`. I'm not sure that it belongs here - is there a general utils module somewhere in the project?

hjoliver commented:
> Temporary PR to check fix on codacy.

Note you can get codacy to analyze your own fork.  Also fine to put up a "work in progress" PR though (which we seem to be labeling with "WIP" now, following @kinow).

hjoliver commented:
(@kinow - that is a cool feature of GitLab).

wxtim commented:
> > Does anyone have any thoughts on how (or whether) to add testing for this bug and solution?
> 
> Maybe add some debug output on the number of jobs submitted, e.g. "50 jobs submitted, batch 1/4", and search for that in test suite output.
> 
> Maybe also make batch size a site config item, so we can scale it down for test suites.

We had a discussion about this: I wanted to make it a config item but other people were not so sure it was necessary.

wxtim commented:
> > I created a new function as a sub-function ...  is there a general utils module somewhere in the project?
> 
> No, but there's several more specific utility modules (with "util" in their filenames). We don't generally bother unless/until functions are used in multiple places.

I pretty much got this response from @oliver-sanders review - alongside the suggestion that I collapse it into a generator comprehension for brevity.

wxtim commented:
## Current State of Play with pull request:
1. I've moved the sorting of the `itasks` list outside the part of the code I've changed so that the list gets sorted before being chopped up.
1. I've changed the generator comprehension back into a list comprehension to facilitate using `len(itasks_batchs)`.
1. I've added a message to the `LOG` for debug mode that can be used by a test to check that the chunking has worked correctly.
1. I've changed the size of the chunks back to 100. I have no idea why I changed it to 50.
1. I've added a simple test designed to ensure that the batching is done correctly. Thanks to @matthewrmshin  for the suite used to test the issue.

hjoliver commented on lib/cylc/task_job_mgr.py:
```diff
@@ -285,33 +285,45 @@ def submit_task_jobs(self, suite, itasks, is_simulation=False):
+            for i, itasks_batch in enumerate(itasks_batches):
+                debug_msg = ("Looking at itasks_batch number {0}/{2} which "
+                             "contains {1} tasks")
+                LOG.debug(debug_msg.format(i + 1,
```
Maybe simplify this message to something like `submitting jobs batch 1/3 (100 tasks)`

matthewrmshin replied:
May actually be more informative if the message includes the command. E.g. insert just before and outside of this loop:

```python
LOG.debug(
    '%s ... # will invoke in batches, sizes=%s',
    cmd, [len(b) for b in itask_batches])
```

(You will have one very informative debug message.)

matthewrmshin commented on lib/cylc/task_job_mgr.py:
```diff
@@ -285,33 +285,45 @@ def submit_task_jobs(self, suite, itasks, is_simulation=False):
+                self.proc_pool.put_command(
+                    SubProcContext(
+                        self.JOBS_SUBMIT,
+                        cmd,
```
Instead of line 290, line 317 and this line, which effectively does

```python
cmd_base = cmd
# ...
cmd = cmd_base + job_log_dirs
# ...
cmd,  # referenced here
```

We can remove line 290 and line 317, and just have `cmd + job_log_dirs` here.

wxtim replied:
Done.

matthewrmshin commented on lib/cylc/task_job_mgr.py:
```diff
@@ -285,33 +285,45 @@ def submit_task_jobs(self, suite, itasks, is_simulation=False):
+            cmd_base = cmd
+            itasks = sorted(itasks, key=lambda itask: itask.identity)
+            itasks_batches = [
+                itasks[i:i + 100] for i in range(0, len(itasks), 100)]
```
Maybe nicer to equalise batch size. E.g. between 101-200 tasks, we run 2 equal size batches, between 201-300 tasks, we run 3 batches, and so on.

wxtim replied:
I've done this, although I'm not sure it's really necessary.

matthewrmshin commented on lib/cylc/task_job_mgr.py:
```diff
@@ -285,33 +285,45 @@ def submit_task_jobs(self, suite, itasks, is_simulation=False):
+                debug_msg = "Submitting jobs batch {0}/{1} ({2} task(s))"
+                LOG.debug(debug_msg.format(i + 1,
+                                           len(itasks_batches),
+                                           len(itasks_batch)))
```
I think my point above still holds true - it is more useful to have a debug statement that includes the leading portion of the command.

It is also best to call the `.debug` method with format string with the format operator and a list a arguments - as intended by the default usage of the method.
See also: https://docs.python.org/3/library/logging.html#logging.Logger.debug

matthewrmshin commented on tests/job-submission/18-check-chunking.t:
```diff
@@ -0,0 +1,52 @@
+
+TEST_NAME=${TEST_NAME_BASE}-itasks-msg
+grep_ok "Submitting jobs batch 1/3 (68 task(s))" $LOG_FILE
+grep_ok "Submitting jobs batch 3/3 (66 task(s))" $LOG_FILE
```
Do you want to print 2/2 as well?

wxtim replied:
I did not think that it was needed.
But I can do if you feel I should.

matthewrmshin commented on tests/job-submission/18-check-chunking.t:
```diff
@@ -0,0 +1,51 @@
+TEST_NAME=${TEST_NAME_BASE}-run
+run_ok $TEST_NAME cylc run --debug --no-detach $SUITE_NAME
+LOG_FILE=$(cylc cat-log $SUITE_NAME -m p)
+echo $LOG_FILE >&2
```
Left over debug echo?

wxtim replied:
oops. :leopard:
